Add coders for std::array and WTF::EnumeratedArray

heycam opened the pull request:
#### 10f809a2fa905f2dfc618bc02de7684113b22084
<pre>
Add coders for std::array and WTF::EnumeratedArray
<a href="https://bugs.webkit.org/show_bug.cgi?id=246898">https://bugs.webkit.org/show_bug.cgi?id=246898</a>
&lt;rdar://problem/101453146&gt;

Reviewed by Myles C. Maxfield.

* Source/WTF/wtf/EnumeratedArray.h:
* Source/WTF/wtf/Forward.h:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: <a href="https://commits.webkit.org/255901@main">https://commits.webkit.org/255901@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9038085a682ec0893b032f80c862fe310f87597

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103482 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163815 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3058 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31277 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86175 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99487 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99517 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2179 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80273 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29208 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84105 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83810 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72160 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85101 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37673 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17637 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80230 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35537 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18899 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27667 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4072 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39416 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41469 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82879 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41351 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38129 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18726 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WTF/wtf/EnumeratedArray.h:
```diff
@@ -43,14 +43,7 @@ class EnumeratedArray {
     using const_pointer = const value_type*;
-
-private:
-    // We add 1 to the size because we expect that LastValue is the maximum value of the enum,
-    // rather than the count of items in the enum.
-    // We're assuming the values in the enum are zero-indexed.
```
Is this comment not useful?

heycam replied:
The fact that we assume the values are zero-based, and that LastValue is indeed the last value and not one past the last value, are already mentioned in the comment at the top of the class, a few lines up. I think the explanation of why we need to add 1 is reasonably self-evident, but I'm happy to leave it in if you prefer.

litherum replied:
Up to you. I'm the author of that comment so I think I'm biased :P

litherum commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -499,6 +499,61 @@ template<typename KeyType, typename ValueType> struct ArgumentCoder<WTF::KeyValu
+    {
+        std::array<std::optional<T>, size> items;
+
+        for (auto& item : items) {
```
Hey, that's nifty. The caller indicates how many items they expect, so there isn't a need to send the size across the wire.

heycam replied:
Yes, it's encoded in the type itself.

heycam committed:
Add coders for std::array and WTF::EnumeratedArray
https://bugs.webkit.org/show_bug.cgi?id=246898
<rdar://problem/101453146>

Reviewed by Myles C. Maxfield.

* Source/WTF/wtf/EnumeratedArray.h:
* Source/WTF/wtf/Forward.h:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: https://commits.webkit.org/255901@main
